File: Problems/0600_0699/0637_Average_of_Levels_in_Binary_Tree/Project_CS/Average_of_Levels_in_Binary_Tree.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution
{
    public List<double> AverageOfLevels_2(TreeNode root)
    {
        List<double> avgs = new List<double>();

        if (root == null)
            return avgs;

        Queue<TreeNode> parentQ = new Queue<TreeNode>();
        Queue<TreeNode> childQ = new Queue<TreeNode>();
        double sum = 0.0;
        
        parentQ.Enqueue(root);
        avgs.Add(System.Convert.ToDouble(root.val));
        
        while (parentQ.Count != 0)
        {
            TreeNode node = parentQ.Dequeue();
            
            if (node.left != null)
            {
                childQ.Enqueue(node.left);
                sum += node.left.val;
            }
            if (node.right != null)
            {
                childQ.Enqueue(node.right);
                sum += node.right.val;
            }
            
            if (parentQ.Count == 0)
            {
                parentQ = childQ;
                childQ = new Queue<TreeNode>();
                
                if(parentQ.Count != 0)
                    avgs.Add(sum / parentQ.Count);
                sum = 0.0;
            }
        }
        
        return avgs;
    }

    public List<double> AverageOfLevels(TreeNode root)
    {
        List<List<int>> data = LevelOrderBottom(root);
        List<double> means = new List<double>();

        for (int i = data.Count - 1; i >= 0; --i)
        {
            means.Add(data[i].Average());
        }
        
        return means;
    }

    public List<List<int>> LevelOrderBottom(TreeNode root)
    {
        var result = new List<List<int>>();
        if (root == null)
            return result;
        
        var queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        
        while(true)
        {
            int nodeCount = queue.Count;
            if(nodeCount == 0)
            {
                break;
            }
            
            var subList = new List<int>();
            while(nodeCount > 0)
            {
                var dataNode = (TreeNode)queue.Dequeue();
                
                subList.Add(dataNode.val);
                
                if(dataNode.left != null)
                {
                    queue.Enqueue(dataNode.left);
                }
                
                if(dataNode.right != null)
                {
                    queue.Enqueue(dataNode.right);
                }
                
                nodeCount--;
            }
            
            result.Insert(0, subList);
        }
        
        return result;
    }

    public void Main(string args)
    {
        Console.WriteLine("args = " + args );
        string flds = args.Replace("\"", "").Replace("[", "").Replace("]", "").Trim();

        OperateTreeNode ope_t = new OperateTreeNode();
        TreeNode root = ope_t.CreateTreeNode(flds);

        Console.Write("root = \n" + ope_t.TreeToStaircaseString(root));
        Console.WriteLine("root = " + ope_t.Tree2str(root));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        List<double> result = AverageOfLevels(root);

        sw.Stop();
        Console.WriteLine("result = [" + String.Join(",", result) + "]");
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}

File: Problems/0600_0699/0637_Average_of_Levels_in_Binary_Tree/Project_Python3/Average_of_Levels_in_Binary_Tree.py
# coding: utf-8

from collections import deque
import os
import sys
import time

from TreeNode.Codec import Codec
from TreeNode.TreeNode import TreeNode
from TreeNode.OperateTreeNode import OperateTreeNode

class Solution:
    def averageOfLevels(self, root: 'TreeNode') -> 'List[float]':
        data = self.levelOrderBottom(root)

        length = len(data)
        mean = [0]*length
        for i in range(length):
            mean[length - 1 - i] = sum(data[i]) / len(data[i])
        
        return mean

    def levelOrderBottom(self, root: 'TreeNode') -> 'List[List[int]]':
        if not root:
            return []
        ans=[[root.val]]
        queue=deque([root])
        while queue:
            cur_val_list=[]
            for i in range(len(queue)):
                root=queue.popleft()
                if root.left:
                    cur_val_list.append(root.left.val)
                    queue.append(root.left)
                if root.right:
                    cur_val_list.append(root.right.val)
                    queue.append(root.right)
            if cur_val_list:
                ans.insert(0,cur_val_list)
        return ans

def main():
    argv = sys.argv
    argc = len(argv)

    if argc < 2:
        print("Usage: python {0} <testdata.txt>".format(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("{0} not found...".format(argv[1]))
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = {0}".format(temp))
        loop_main(temp)
    #   print("Hit Return to continue...")
    #   input()

def loop_main(temp):
    flds = temp.replace(", ", ",").replace("\"", "").replace("[", "").replace("]", "").rstrip()

    ope_t = OperateTreeNode()
    
    if len(flds) > 0:
        root = ope_t.createTreeNode(flds.split(","))
    else:
        root = None

    print("root = \n{0}".format(ope_t.treeToStaircaseString(root)))
    print("root = {0}".format(ope_t.tree2str(root)))

    sl = Solution()
    time0 = time.time()

    result = sl.averageOfLevels(root)

    time1 = time.time()

    print("result = {0}".format(result))
    print("Execute time ... : {0:f}[s]\n".format(time1 - time0))

if __name__ == "__main__":
    main()
